### ISSUE TITLE
Please add support for Ubuntu Push

### ISSUE THREAD

**Issue by ted-gould** [processone/ejabberd#507]
Noticed that ejabberd (at least the commercial edition) supports Android and Apple push notifications. It would be great if Ubuntu Phone push notifications could also be supported.

https://developer.ubuntu.com/en/apps/platform/guides/push-notifications-server-guide/

In the example "data" is filled out, but it could be anything including an XMPP message (which is what I'd guess would be the most useful)

**Comment by royneary** [comment]
I'm implementing XEP-0357 as an ejabberd module.  The XEP does not define the app-server part, but my implementation has an app server included that tries to be generic (at least for instant messaging applications). You can find it on my github page (mod_push). It's not usable yet, but push notifications have already been sent to my Ubuntu phone.

**Comment by ted-gould** [comment]
Oh, that's interesting. My understanding is that 357 allows for creating
a server that is actually independent of the Jabber server itself, just
that the Jabber server needs Pub-Sub capability. Could that module then
be used to connect to other Jabber servers?

Do you know if this is the way that Apple and Android push notifications
are integrated into ejabberd?

**Comment by royneary** [comment]
That's correct, an app server provides a pubsub node at its pubsub service and an XMPP server sends notifications via that service. So the minimum requirement of the XEP is that my module can publish items on a pubsub node.
In addition to that my module implements the app server part. A user can register at this app server using adhoc commands like this:

```xml
<iq type='set' to='montague.net' id='exec1'>
  <command xmlns='http://jabber.org/protocol/commands'
           node='register-push-ubuntu'
           action='execute'/>
</iq>
```

A node on the configured pubsub service will be created and the service's hostname + node name is sent back to the user. She can then send this information to her XMPP server using the enable stanza from XEP-0357. The XMPP server and the app server can be the same instance (same hostname or different hostnames) or different instances.
The idea behind this all-in-one-module approach is to make it possible to  decentralize app servers instead of having one centralized server per app. This is currently only possible with Ubuntu's push service because they don't require authentication using the app developer's credentials like the other services do. Of course for decentralization it would be nice to also have a XEP-0114 component that other servers besides ejabberd can use, but first things first.

I don't know how push is implemented in the commercial edition.

**Comment by bangarharshit** [comment]
@royneary Can we merge mod_push with 15.10 ejabberd tag? Right now I see it's merged to somewhere in between 15.09 and 15.10. 15.09 have some issues related to xml-rpc and it will be great if it can be merged with something stable. I can send a pull request if it's ok.

**Comment by weiss** [comment]
You might want to subsribe issue #1379 instead of this one.
